stratosphere/loader/source/ldr_content_management.cpp

```diff
@@ -13,459 +13,331 @@
-static bool g_mounted_hbl_nsp = false;
-static char g_hbl_sd_path[FS_MAX_PATH+1] = "@Sdcard:/atmosphere/hbl.nsp\x00";
+namespace sts::ldr {
 
-static OverrideKey g_default_override_key = {
-    .key_combination = KEY_L,
-    .override_by_default = true
-};
+    namespace {
 
-struct HblOverrideConfig {
-    OverrideKey override_key;
-    u64 title_id;
-    bool override_any_app;
-};
+        /* DeviceNames. */
+        constexpr const char *CodeFileSystemDeviceName   = "code";
+        constexpr const char *HblFileSystemDeviceName    = "hbl";
+        constexpr const char *SdCardFileSystemDeviceName = "sdmc";
 
-static HblOverrideConfig g_hbl_override_config = {
-    .override_key = {
-        .key_combination = KEY_R,
-        .override_by_default = true
-    },
-    .title_id = TitleId_AppletPhotoViewer,
-    .override_any_app = false
-};
+        constexpr const char *SdCardStorageMountPoint = "@Sdcard";
 
-/* Static buffer for loader.ini contents at runtime. */
-static char g_config_ini_data[0x800];
+        /* Globals. */
+        bool g_has_mounted_sd_card = false;
 
-/* SetExternalContentSource extension */
-static std::map<u64, ContentManagement::ExternalContentSource> g_external_content_sources;
+        ncm::TitleId g_should_override_title_id;
+        bool g_should_override_hbl = false;
+        bool g_should_override_sd  = false;
 
-Result ContentManagement::MountCode(u64 tid, FsStorageId sid) {
-    char path[FS_MAX_PATH] = {0};
-
-    /* We defer SD card mounting, so if relevant ensure it is mounted. */
-    if (!g_has_initialized_fs_dev) {
-        TryMountSdCard();
-    }
-
-    if (g_has_initialized_fs_dev) {
-        RefreshConfigurationData();
-    }
-
-    if (ShouldOverrideContentsWithSD(tid) && R_SUCCEEDED(MountCodeNspOnSd(tid))) {
-        return ResultSuccess;
-    }
-
-    R_TRY(ResolveContentPath(path, tid, sid));
-
-    /* Fix up path. */
-    for (unsigned int i = 0; i < FS_MAX_PATH && path[i] != '\x00'; i++) {
-        if (path[i] == '\\') {
-            path[i] = '/';
+        /* Helpers. */
+        inline void FixFileSystemPath(char *path) {
+            for (size_t i = 0; i < FS_MAX_PATH && path[i]; i++) {
+                if (path[i] == '\\') {
+                    path[i] = '/';
+                }
+            }
         }
-    }
-
-    /* Always re-initialize fsp-ldr, in case it's closed */
-    DoWithSmSession([&]() {
-        R_ASSERT(fsldrInitialize());
-    });
-    ON_SCOPE_EXIT { fsldrExit(); };
 
-    R_TRY(fsldrOpenCodeFileSystem(tid, path, &g_CodeFileSystem));
+        inline const char *GetRelativePathStart(const char *relative_path) {
+            while (*relative_path == '/' || *relative_path == '\\') {
+                relative_path++;
+            }
+            return relative_path;
+        }
 
-    fsdevMountDevice("code", g_CodeFileSystem);
-    TryMountHblNspOnSd();
-    return ResultSuccess;
-}
+        void UpdateShouldOverrideCache(ncm::TitleId title_id) {
```

Comment:
"Sampled only once per launch" -- not possible, because launches can be asynchronous with regards to the commands sent to perform them. The cache should properly get us down to the minimum of two samples.

We could maybe get it down to one by using an even heavier cache (std::unordered_map?)

Reply:
Oh, you mean between `GetProgramInfo` and `CreateProcess`? That does make things hard... but since we control PM too, perhaps solve both problems at the same time by having PM request `ams.mitm` to sample button state before launch, then have loader read out what `ams.mitm` had sampled.

Reply:
Yeah, that actually might be a good idea. Stratosphere namespacing would let us actually provide services inside ams.mitm sanely. I'll be rewriting PM and ams.mitm soon -- I'll see about doing something like that. I'll leave this as is, for now?

Reply:
Sounds good.